gcc/d/expr.cc

```diff
@@ -2625,6 +2625,7 @@ class ExprVisitor : public Visitor
 	/* Array type string length includes the null terminator.  */
 	dinteger_t length = (e->len * e->sz) + 1;
 	tree value = build_string (length, string);
+	((char*)TREE_STRING_POINTER (value))[length - 1] = '\0';
```

Comment:
@ibuclaw The original code above passes string-length+1 to `build_string`. Even though the docs say the length passed should include the zero terminator, `build_string` internally allocates one byte more and adds a zero separator as the last byte. So `build_string(20, string)` allocates 21 bytes, copies 20 bytes from `string` and sets byte 21 to zero.

However, it seems in DDMD, `string` is not guaranteed to be zero terminated. There was one case in `std.net.curl` where this was not the case. Combined with the `build_string` behavior and us passing `length+1` to `build_string`, this led to 1 byte of garbage at the end of the string.
This then caused the `std.net.curl` test to fail loading one function using `dlsym` as the name contained that garbage byte and as the test seems not prepared to fail it somehow manages to cause a deadlock ultimately leading to a test timeout.

I've added a workaround for testing in this commit. Is there any reason we can't simply pass the original length to `build_string` and rely on that function to zero-terminate? I guess the only other proper solution is manually copying and zero-terminating the `string` before calling `build_string` or is there some clever solution for this?

Reply:
It's because of https://bugzilla.gdcproject.org/show_bug.cgi?id=285

I just checked by reverting it, and the bug still the case, perhaps this should instead be fixed in the frontend itself to ensure null termination?  I see the last value is `-1`, seems like the default `char` initializer.

Reply:
Actually, this lacking of null terminator was caused by https://github.com/dlang/dmd/pull/5453, which removed a bunch of setting them.